chainer/functions/array/as_strided.py

```diff
@@ -0,0 +1,429 @@
+
+    Args:
+        shape (tuple of int): The shape of output.
+        strides (tuple of int):
+            The strides of output, given in the unit of steps.
+        storage_offset (int):
+            The offset between the head of allocated memory and the pointer of
+            first element, given in the unit of steps.
+
+    Returns:
+        int: The leftest pointer in the array
+    """
+    ndim = len(shape)
+    shape_negative = [shape[i] for i in range(ndim) if strides[i] < 0]
+    strides_negative = [strides[i] for i in range(ndim) if strides[i] < 0]
+    if len(strides_negative) == 0:
+        return storage_offset
+    else:
+        return storage_offset + reduce(
+            lambda base, shape_strides:
+            base + (shape_strides[0] - 1) * shape_strides[1],
+            zip(shape_negative, strides_negative), 0)
+
+
+def _max_index(shape, strides, storage_offset):
+    """Returns the rightest index in the array
+
+    Args:
+        shape (tuple of int): The shape of output.
+        strides (tuple of int): The strides of output, given in unit-steps.
+        storage_offset (int):
+            The offset between the head of allocated memory and the pointer of
+            first element, given in the unit of steps.
+
+    Returns:
+        int: The rightest pointer in the array
+    """
+    ndim = len(shape)
+    shape_positive = [shape[i] for i in range(ndim) if strides[i] > 0]
+    strides_positive = [strides[i] for i in range(ndim) if strides[i] > 0]
+    if len(strides_positive) == 0:
+        return storage_offset
+    else:
+        return storage_offset + reduce(
+            lambda base, shape_strides:
+            base + (shape_strides[0] - 1) * shape_strides[1],
+            zip(shape_positive, strides_positive), 0)
+
+
+def _index_add(augend, indices, addend):
+    """Wrapper of :func:`cupyx.scatter_add` and :func:`numpy.add.at`
+
+    Args:
+        augend (:class:`numpy.ndarray` or :class:`cupy.ndarray`):
+            The array modified in-place.
+        indices (:class:`numpy.ndarray` or :class:`cupy.ndarray`):
+            The indices of ``augend``. The shape is the same to the ``addend``.
+        addend (:class:`numpy.ndarray` or :class:`cupy.ndarray`):
+            The array to be added.
+
+    Returns:
+        None
+    """
+    if isinstance(augend, cuda.ndarray):
+        cuda.cupyx.scatter_add(augend, indices, addend)
+    elif isinstance(augend, np.ndarray):
+        np.add.at(augend, indices, addend)
+
+
+def _get_base_array(array):
+    """Get the founder of :class:`numpy.ndarray`.
+
+    Args:
+        array (:class:`numpy.ndarray`):
+            The view of the base array.
+
+    Returns:
+        :class:`numpy.ndarray`:
+            The base array.
+    """
+    base_array_candidate = array
+    while base_array_candidate.base is not None:
+        base_array_candidate = base_array_candidate.base
+    return base_array_candidate
+
+
+def _stride_array(array, shape, strides, storage_offset):
```

Comment:
Why do not use `as_strided` ?

Reply:
Do you mean `np.as_strided` and `cp.as_strided` (which to be merged in the next release)?

There are two reasons:
1. These functions do not detect invalid range (i.e. out_of_range exception)
2. These functions do not support storage_offset, which is supported in pytorch and enables the users to create flipped view.

Reply:
OK, I understood.